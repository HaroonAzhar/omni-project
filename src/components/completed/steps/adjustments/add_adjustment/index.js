import React, { useRef } from "react";
import { Field, Form } from "react-final-form";

import {
  Button,
  H2,
  PriceField,
  SelectInput,
  TextAreaInput,
  TextInput,
} from "components/atoms";
import { formValidation } from "utils";
import { saveAdjustment } from "utils/requests";

import getTransactionTypeOptions from "./get_transaction_type_options";
import validationSchema from "./validation_schema";
import useSaveAdjustment from "./use_save_adjustment";
import { AddContainer } from "../../shared_styles/styled_add";
import mapTransactionToBalance from "./map_transaction_to_balance";
import useAdjustmentsData from "../view_adjustments/use_adjustments_data";

const AddAdjustment = () => {
  const validate = async (values) => formValidation(validationSchema, values);

  const { fetchAdjustmentsAndStore } = useAdjustmentsData(false);
  const saveNewAdjustment = useSaveAdjustment(saveAdjustment);
  const formRef = useRef();

  const submit = (values) => {
    const sendData = {
      ...values,
      ...mapTransactionToBalance(values.TransactionType),
    };
    saveNewAdjustment(sendData).then((res) => {
      if (res) {
        fetchAdjustmentsAndStore();
        formRef.current.reset();
      }
    });
  };
  return (
    <AddContainer>
      <H2>Add Adjustment</H2>
      <Form
        onSubmit={submit}
        validate={validate}
        render={({ handleSubmit, submitting, form }) => {
          formRef.current = form;
          return (
            <form onSubmit={handleSubmit}>
              <Field
                component={SelectInput}
                name="TransactionType"
                label="Transaction Type"
                type="text"
                optionsWithGroups={getTransactionTypeOptions()}
              />
              <Field
                component={TextInput}
                name="ActualDate"
                label="Actual Date"
                type="date"
              />
              <PriceField label="Amount" name="Amount" />
              <Field
                component={TextAreaInput}
                name="Description"
                label="Description"
                type="text"
              />
              <Field
                component={TextAreaInput}
                name="InternalNote"
                label="Internal Notes"
                type="text"
              />
              <Button type="submit" disabled={submitting}>
                Add
              </Button>
            </form>
          );
        }}
      />
    </AddContainer>
  );
};

export default AddAdjustment;
