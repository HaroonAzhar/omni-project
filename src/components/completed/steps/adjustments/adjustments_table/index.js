/* eslint-disable react/jsx-props-no-spreading */
import React from "react";
import { useTable, useSortBy } from "react-table";
import PropTypes from "prop-types";

import { StyledSortingArrow } from "components/molecules/table/styled_table";

import { StyledTable } from "./styled_adjustments_table";
const AdjustmentTable = ({ columns, adjustments, sortBy = [] }) => {
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows: AdjustmentsRows,
    prepareRow,
  } = useTable(
    {
      columns,
      data: adjustments,
      initialState: { sortBy },
    },
    useSortBy
  );
  const renderRevisions = (revisions) => {
    return revisions.map((revision) => {
      return <p>{revision}</p>;
    });
  };

  return (
    <>
      <StyledTable>
        <table {...getTableProps()}>
          <thead>
            {headerGroups.map((headerGroup) => (
              <tr {...headerGroup.getHeaderGroupProps()}>
                {headerGroup.headers.map(
                  ({
                    getHeaderProps,
                    render,
                    getSortByToggleProps,
                    isSorted,
                    isSortedDesc,
                    canSort,
                  }) => (
                    <th {...getHeaderProps(canSort && getSortByToggleProps())}>
                      {render("Header")}
                      {canSort && (
                        <StyledSortingArrow
                          isActive={isSorted}
                          isAscending={isSortedDesc}
                        />
                      )}
                    </th>
                  )
                )}
              </tr>
            ))}
          </thead>

          <tbody {...getTableBodyProps()}>
            {AdjustmentsRows.map((row) => {
              prepareRow(row);
              return (
                <React.Fragment {...row.getRowProps()}>
                  <tr>
                    {row.cells.map((cell) => {
                      return (
                        <td {...cell.getCellProps()}>{cell.render("Cell")}</td>
                      );
                    })}
                  </tr>

                  {row.original.subRowsData.internalNote && (
                    <tr>
                      <td
                        className="internal-notes-sub-row"
                        colSpan={row.cells.length}
                      >
                        <p>{row.original.subRowsData.internalNote} </p>
                      </td>
                    </tr>
                  )}

                  {row.original.subRowsData.revisions && (
                    <tr>
                      <td
                        className="revisions-sub-row"
                        colSpan={row.cells.length}
                      >
                        <p> Revisions</p>
                        {renderRevisions(row.original.subRowsData.revisions)}
                      </td>
                    </tr>
                  )}
                </React.Fragment>
              );
            })}
          </tbody>
        </table>
      </StyledTable>
    </>
  );
};

AdjustmentTable.propTypes = {
  sortBy: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      desc: PropTypes.bool,
    })
  ),
  sortable: PropTypes.bool,
  columns: PropTypes.array,
  adjustments: PropTypes.array,
};

export default AdjustmentTable;
