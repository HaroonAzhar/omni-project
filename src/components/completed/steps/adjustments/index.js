import React from "react";
import { titleize } from "inflected";

import { H1 } from "components/atoms";

import { ADJUSTMENTS } from "../../completed_steps";
import AddAdjustment from "./add_adjustment";
import ViewAdjustments from "./view_adjustments";
import {
  Background,
  ContentWrapper,
  StepContainer,
  TitleWrapper,
} from "../shared_styles";
import FilterModal from "./filter_adjustments";

const Adjustments = () => {
  return (
    <Background>
      <ContentWrapper>
        <TitleWrapper>
          <H1>{titleize(ADJUSTMENTS)}</H1>
          <FilterModal />
        </TitleWrapper>
        <StepContainer>
          <ViewAdjustments />
          <AddAdjustment />
        </StepContainer>
      </ContentWrapper>
    </Background>
  );
};
export default Adjustments;
