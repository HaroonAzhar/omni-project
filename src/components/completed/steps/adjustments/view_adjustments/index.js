import React from "react";

import columns from "./columns";
import useAdjustmentsData from "./use_adjustments_data";
import AdjusmentTable from "../adjustments_table";
import mapAdjustmentsForTable from "./map_adjustments_for_table";

const ViewAdjustments = () => {
  const { adjustments } = useAdjustmentsData(true);

  const adjustmentsData = mapAdjustmentsForTable(adjustments);
  return (
    <AdjusmentTable
      className="view-adjustments-table"
      columns={columns}
      adjustments={adjustmentsData}
    />
  );
};

export default ViewAdjustments;
