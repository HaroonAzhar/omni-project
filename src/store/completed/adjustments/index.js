import { createSlice } from "@reduxjs/toolkit";

const adjustmentsSlice = createSlice({
  name: "adjustments",
  initialState: [],
  reducers: {
    insertAdjustmentsData(state, action) {
      const { adjustments } = action.payload;
      return [...adjustments];
    },
  },
});

export const { insertAdjustmentsData } = adjustmentsSlice.actions;

export default adjustmentsSlice.reducer;
