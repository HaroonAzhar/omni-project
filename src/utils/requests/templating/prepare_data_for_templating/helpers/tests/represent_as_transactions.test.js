import statementResults from "components/completed/steps/cashflows/utils/__tests__/data/statement_results";
import adjustments from "components/completed/steps/cashflows/utils/__tests__/data/adjustments";

import representAsTransactions from "../represent_as_transactions";

describe("representAsTransactions", () => {
  it("e2e test", () => {
    const actualTransactions = representAsTransactions({
      adjustments,
      statementResults,
    });

    expect(actualTransactions).toMatchSnapshot();
  });
});
