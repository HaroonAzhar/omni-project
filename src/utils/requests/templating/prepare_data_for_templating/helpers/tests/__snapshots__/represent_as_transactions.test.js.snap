// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`representAsTransactions e2e test 1`] = `
Array [
  Object {
    "Date": "2019-12-31T00:00:00",
    "Debits": 3850,
    "Description": " Interest",
    "FormattedDate": "31/12/2019 to 30/01/2020",
    "Type": "Interest",
  },
  Object {
    "Date": "2020-01-31T00:00:00",
    "Debits": 3850,
    "Description": " Interest",
    "FormattedDate": "31/01/2020 to 28/02/2020",
    "Type": "Interest",
  },
  Object {
    "Date": "2020-02-29T00:00:00",
    "Debits": 3850,
    "Description": " Interest",
    "FormattedDate": "29/02/2020 to 30/03/2020",
    "Type": "Interest",
  },
  Object {
    "Date": "2020-03-31T00:00:00",
    "Debits": 3850,
    "Description": " Interest",
    "FormattedDate": "31/03/2020 to 29/04/2020",
    "Type": "Interest",
  },
  Object {
    "Date": "2020-04-30T00:00:00",
    "Debits": 3850,
    "Description": " Interest",
    "FormattedDate": "30/04/2020 to 30/05/2020",
    "Type": "Interest",
  },
  Object {
    "Date": "2020-05-31T00:00:00",
    "Debits": 3850,
    "Description": " Interest",
    "FormattedDate": "31/05/2020 to 29/06/2020",
    "Type": "Interest",
  },
  Object {
    "Date": "2020-06-30T00:00:00",
    "Debits": 3850,
    "Description": " Interest",
    "FormattedDate": "30/06/2020 to 30/07/2020",
    "Type": "Interest",
  },
  Object {
    "Date": "2020-07-31T00:00:00",
    "Debits": 3850,
    "Description": " Interest",
    "FormattedDate": "31/07/2020 to 30/08/2020",
    "Type": "Interest",
  },
  Object {
    "Date": "2020-08-31T00:00:00",
    "Debits": 3850,
    "Description": " Interest",
    "FormattedDate": "31/08/2020 to 29/09/2020",
    "Type": "Interest",
  },
  Object {
    "Date": "2020-09-30T00:00:00",
    "Debits": 3850,
    "Description": " Interest",
    "FormattedDate": "30/09/2020 to 30/10/2020",
    "Type": "Interest",
  },
  Object {
    "Date": "2020-10-31T00:00:00",
    "Debits": 3850,
    "Description": " Interest",
    "FormattedDate": "31/10/2020 to 29/11/2020",
    "Type": "Interest",
  },
  Object {
    "Date": "2020-11-30T00:00:00",
    "Debits": 3593.3333333333335,
    "Description": " Interest",
    "FormattedDate": "30/11/2020 to 27/12/2020",
    "Type": "Interest",
  },
  Object {
    "Date": "2020-12-28T00:00:00",
    "Debits": 478.90614896459584,
    "Description": " Interest",
    "FormattedDate": "28/12/2020 to 29/12/2020",
    "Type": "Interest",
  },
  Object {
    "Date": "2020-12-31T00:00:00",
    "Debits": 350.57281563126253,
    "Description": "Extension Interest",
    "FormattedDate": "31/12/2020 to 31/12/2020",
    "Type": "Interest",
  },
  Object {
    "Date": "2021-01-01T00:00:00",
    "Debits": 350.57281563126253,
    "Description": "Extension Interest",
    "FormattedDate": "01/01/2021 to 01/01/2021",
    "Type": "Interest",
  },
  Object {
    "Date": "2021-01-02T00:00:00",
    "Debits": 350.57281563126253,
    "Description": "Extension Interest",
    "FormattedDate": "02/01/2021 to 02/01/2021",
    "Type": "Interest",
  },
  Object {
    "Date": "2021-01-03T00:00:00",
    "Debits": 9114.893206412826,
    "Description": "Extension Interest",
    "FormattedDate": "03/01/2021 to 28/01/2021",
    "Type": "Interest",
  },
  Object {
    "Date": "2021-01-29T00:00:00",
    "Debits": 350.57281563126253,
    "Description": "Extension Interest",
    "FormattedDate": "29/01/2021 to 29/01/2021",
    "Type": "Interest",
  },
  Object {
    "Date": "2021-01-31T00:00:00",
    "Debits": 721.4788479024716,
    "Description": "Default Interest",
    "FormattedDate": "31/01/2021 to 01/02/2021",
    "Type": "Interest",
  },
  Object {
    "Credits": 1000,
    "Date": "2021-02-02T00:00:00",
    "Description": "Interest Rebate (Manual)",
    "FormattedDate": "02/02/2021",
    "Type": "Adjustment",
  },
  Object {
    "Date": "2021-02-02T00:00:00",
    "Debits": 1000,
    "Description": "Further Advance on a Single Advance Loan",
    "FormattedDate": "02/02/2021",
    "Type": "Adjustment",
  },
  Object {
    "Date": "2021-02-02T00:00:00",
    "Debits": 360.7394239512358,
    "Description": "Default Interest",
    "FormattedDate": "02/02/2021 to 02/02/2021",
    "Type": "Interest",
  },
  Object {
    "Credits": 1000,
    "Date": "2021-02-03T00:00:00",
    "Description": "Ad-Hoc Balance Reduction",
    "FormattedDate": "03/02/2021",
    "Type": "Adjustment",
  },
  Object {
    "Credits": 123,
    "Date": "2021-02-03T00:00:00",
    "Description": "Import Adjustment",
    "FormattedDate": "03/02/2021",
    "Type": "Adjustment",
  },
  Object {
    "Credits": 123,
    "Date": "2021-02-03T00:00:00",
    "Description": "Ad-Hoc Balance Reduction",
    "FormattedDate": "03/02/2021",
    "Type": "Adjustment",
  },
  Object {
    "Credits": 1000,
    "Date": "2021-02-03T00:00:00",
    "Description": "Import Adjustment",
    "FormattedDate": "03/02/2021",
    "Type": "Adjustment",
  },
  Object {
    "Date": "2021-02-03T00:00:00",
    "Debits": 1000,
    "Description": "Import Adjustment",
    "FormattedDate": "03/02/2021",
    "Type": "Adjustment",
  },
  Object {
    "Date": "2021-02-03T00:00:00",
    "Debits": 1000,
    "Description": "Litigation",
    "FormattedDate": "03/02/2021",
    "Type": "Adjustment",
  },
  Object {
    "Date": "2021-02-03T00:00:00",
    "Debits": 360.4934239512358,
    "Description": "Default Interest",
    "FormattedDate": "03/02/2021 to 03/02/2021",
    "Type": "Interest",
  },
  Object {
    "Credits": 1000,
    "Date": "2021-02-04T00:00:00",
    "Description": "Import Adjustment",
    "FormattedDate": "04/02/2021",
    "Type": "Adjustment",
  },
  Object {
    "Date": "2021-02-04T00:00:00",
    "Debits": 2156.9605437074147,
    "Description": "Default Interest",
    "FormattedDate": "04/02/2021 to 09/02/2021",
    "Type": "Interest",
  },
  Object {
    "Credits": 1000,
    "Date": "2021-02-10T00:00:00",
    "Description": "Partial Redemption - Foo",
    "FormattedDate": "10/02/2021",
    "Type": "Adjustment",
  },
  Object {
    "Credits": 1000,
    "Date": "2021-02-10T00:00:00",
    "Description": "Redemption",
    "FormattedDate": "10/02/2021",
    "Type": "Adjustment",
  },
  Object {
    "Date": "2021-02-10T00:00:00",
    "Debits": 1000,
    "Description": "Ad-Hoc Cost",
    "FormattedDate": "10/02/2021",
    "Type": "Adjustment",
  },
  Object {
    "Date": "2021-02-10T00:00:00",
    "Debits": 358.4934239512358,
    "Description": "Default Interest",
    "FormattedDate": "10/02/2021 to 10/02/2021",
    "Type": "Interest",
  },
  Object {
    "Date": "2021-02-11T00:00:00",
    "Debits": 123,
    "Description": "Import Adjustment",
    "FormattedDate": "11/02/2021",
    "Type": "Adjustment",
  },
  Object {
    "Credits": 1000,
    "Date": "2021-02-11T00:00:00",
    "Description": "Partial Redemption",
    "FormattedDate": "11/02/2021",
    "Type": "Adjustment",
  },
  Object {
    "Date": "2021-02-11T00:00:00",
    "Debits": 357.6164239512358,
    "Description": "Default Interest",
    "FormattedDate": "11/02/2021 to 11/02/2021",
    "Type": "Interest",
  },
  Object {
    "Date": "2021-02-12T00:00:00",
    "Debits": 1000,
    "Description": "Ad-Hoc Cost",
    "FormattedDate": "12/02/2021",
    "Type": "Adjustment",
  },
  Object {
    "Date": "2021-02-12T00:00:00",
    "Debits": 1075.8492718537075,
    "Description": "Default Interest",
    "FormattedDate": "12/02/2021 to 14/02/2021",
    "Type": "Interest",
  },
]
`;
